### PR TITLE
DEI-185: Added workflow to publish PyPi package to Artifactory and PyPi

### PR DESCRIPTION
**Pull request by aslotte**
## JIRA ticket
DEI-185

## Description

The publish PyPi package action does the following
- Validates tag and branch
- Awaits approval from reviewer (defined PROD environment) before proceeding
- Publishes package to Artifactory
- Publishes package to PyPi

Updated the `code_qulity` action with
- Resolve packages from JFrog Artifactory for `poetry`
- Resolve packages from JFrog Artifactory for `pipenv`

## Automated testing

- [ ] Unit tests
- [ ] Integration tests
- [ ] API tests
- [ ] End-To-End tests
- [x] N/A (please clarify)

### CODE REVIEW

File: .github/workflows/python_code_quality.yml [hunk 1]
```diff
@@ -85,6 +91,9 @@ jobs:
           python-version: ${{ inputs.python-version }}
       # if there's a Pipfile, use pipenv
       - name: Test with pipenv
+        env: 
+          JFROG_ARTIFACTORY_USER: ${{ secrets.JFROG_ARTIFACTORY_USER }}
+          JFROG_ARTIFACTORY_PASSWORD: ${{ secrets.JFROG_ARTIFACTORY_PASSWORD }}      
         if: hashFiles('Pipfile') != ''
```
Comment (VladimirShteyn) [inline .github/workflows/python_code_quality.yml:97]: This isn't sufficient for getting `pipenv` to install from a private PyPi. [See here for instructions](https://pipenv-fork.readthedocs.io/en/latest/advanced.html#injecting-credentials-into-pipfiles-via-environment-variables). The summary is, unfortunately, to be able to install from a private PyPi, `pipenv` requires adding the password to `Pipfile`. Perhaps we should use something like [aws-labs git-secrets](https://github.com/awslabs/git-secrets#example-walkthrough)?

File: .github/workflows/publish_pypi_package.yml [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        ts-sdk init -o org -p test-protocol -t ts-slug -f ./new-protocol
+        pushd .
+        cd ./new-protocol/task-script
+        pipenv install --dev ../../dist/*.whl
+        pipenv install --dev
+        pipenv run python -m pytest -s
+        popd
+
+    - name: Build package
+      if: hashFiles('poetry.lock') != ''
```
Comment (VladimirShteyn) [inline .github/workflows/publish_pypi_package.yml:98]: Should we raise an error if there's no `poetry.lock` file?

File: .github/workflows/publish_pypi_package.yml [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      with:
+        python-version: ${{ inputs.python-version }}
+
+    # Temporary - to be removed once ts-sdk uses poetry
+    - name: Build ts-sdk package
+      if: hashFiles('poetry.lock') == ''    
+      run: |
+        pip install --upgrade pip setuptools wheel
+        python3 setup.py bdist_wheel
+        pip install ./dist/*.whl
```
Comment (VladimirShteyn) [inline .github/workflows/publish_pypi_package.yml:87]: Lines 87-95 are a smoke test; they don't build the package. The test is already run as part of CI so these lines can be safely deleted here.

**Review by VladimirShteyn — CHANGES_REQUESTED** [review]
Request Changes because the `python_code_quality.yml` worfklow isn't compatible with `pipenv`.

File: .github/workflows/publish_pypi_package.yml [hunk 4]
```diff
@@ -0,0 +1,100 @@
+name: Publish PyPi package
+on:
+  workflow_call:
+    inputs:
+      python-version:
+        type: string
+        default: "3.7"
+        required: false
+    secrets:
+      JFROG_ARTIFACTORY_USER:
```
[the remark this answers is not in the document]
Comment (aslotte) [reply .github/workflows/publish_pypi_package.yml:10]: They don't technically need to be secrets, but having them as secrets makes it easier to change them if needed across all repos.

File: .github/workflows/python_code_quality.yml (hunk 1 above)
Comment (aslotte) [reply .github/workflows/python_code_quality.yml:97]: That's correct, by itself it's not sufficient, but the plan is to add the following to each `Pipfile` and that will make it work (tested it locally)

```
[[source]]
name = "ts_pypi_virtual"
url = "https://${JFROG_ARTIFACTORY_USER}:${JFROG_ARTIFACTORY_PASSWORD}@tetrascience.jfrog.io/artifactory/api/pypi/ts-pypi-virtual/simple"
verify_ssl = true
```

File: .github/workflows/publish_pypi_package.yml (hunk 2 above)
Comment (aslotte) [reply .github/workflows/publish_pypi_package.yml:98]: No need, the workflow will fail by itself

Comment (aslotte) [reply .github/workflows/publish_pypi_package.yml:98]: We also have a limited number of packages we are publishing

File: .github/workflows/publish_pypi_package.yml (hunk 3 above)
Comment (aslotte) [reply .github/workflows/publish_pypi_package.yml:87]: Awesome, I'll remove them!

File: .github/workflows/publish_pypi_package.yml (hunk 2 above)
Comment (VladimirShteyn) [reply .github/workflows/publish_pypi_package.yml:92]: > No need, the workflow will fail by itself

I don't wanna hold up the PR too much, but as the Workflow currently stands the error message would not be very clear. I recommend adding a test that raises an error like, "Only publishing from `poetry` is implemented. No poetry packaging metadata found."

File: .github/workflows/python_code_quality.yml (hunk 1 above)
Comment (VladimirShteyn) [reply .github/workflows/python_code_quality.yml:97]: We should add this to the [Task Script cookiecutter template](https://github.com/tetrascience/ts-cookiecutter/blob/main/task-script-repo/%7B%7B%20cookiecutter.local_repo_name%20%7D%7D/Pipfile) (easy) and to the [migration script](https://github.com/tetrascience/ts-version-folders-to-git) (more complicated).

File: .github/workflows/publish_pypi_package.yml (hunk 4 above)
[the remark this answers is not in the document]
Comment (VladimirShteyn) [reply .github/workflows/publish_pypi_package.yml:10]: I recommend adding a comment explaining that.

File: .github/workflows/python_code_quality.yml (hunk 1 above)
Comment (VladimirShteyn) [reply .github/workflows/python_code_quality.yml:97]: ☝️ implementing the changes in other repos shouldn't hold up this PR though.

File: .github/workflows/publish_pypi_package.yml (hunk 2 above)
Comment (VladimirShteyn) [reply .github/workflows/publish_pypi_package.yml:92]: From a Slack conversation, it looks like `setup.py` publishing is handled by the `pypa/gh-action-pypi-publish` below. Approved.

**Review by VladimirShteyn — APPROVED** [review]

**Pull request merged**